pbsmetrics/prometheus/prometheus.go

```diff
@@ -0,0 +1,242 @@
+		Buckets:   buckets,
+	}
+	return prometheus.NewHistogramVec(opts, labels)
+}
+
+func (me *Metrics) RecordConnectionAccept(success bool) {
+	if success {
+		me.connCounter.Inc()
+	} else {
+		me.connError.WithLabelValues("accept_error").Inc()
+	}
+
+}
+
+func (me *Metrics) RecordConnectionClose(success bool) {
+	if success {
+		me.connCounter.Dec()
+	} else {
+		me.connError.WithLabelValues("close_error").Inc()
+	}
+}
+
+func (me *Metrics) RecordRequest(labels pbsmetrics.Labels) {
+	me.requests.With(resolveLabels(labels)).Inc()
+}
+
+func (me *Metrics) RecordImps(labels pbsmetrics.Labels, numImps int) {
+	me.imps.With(resolveLabels(labels)).Add(float64(numImps))
+}
+
+func (me *Metrics) RecordRequestTime(labels pbsmetrics.Labels, length time.Duration) {
+	time := float64(length) / float64(time.Second)
+	me.reqTimer.With(resolveLabels(labels)).Observe(time)
+}
+
+func (me *Metrics) RecordAdapterRequest(labels pbsmetrics.AdapterLabels) {
+	me.adaptRequests.With(resolveAdapterLabels(labels)).Inc()
+}
+
+func (me *Metrics) RecordAdapterBidReceived(labels pbsmetrics.AdapterLabels, bidType openrtb_ext.BidType, hasAdm bool) {
+	me.adaptBids.With(resolveBidLabels(labels, bidType, hasAdm)).Inc()
+}
+
+func (me *Metrics) RecordAdapterPrice(labels pbsmetrics.AdapterLabels, cpm float64) {
+	me.adaptPrices.With(resolveAdapterLabels(labels)).Observe(cpm)
+}
+
+func (me *Metrics) RecordAdapterTime(labels pbsmetrics.AdapterLabels, length time.Duration) {
+	time := float64(length) / float64(time.Second)
+	me.adaptTimer.With(resolveAdapterLabels(labels)).Observe(time)
+}
+
+func (me *Metrics) RecordCookieSync(labels pbsmetrics.Labels) {
+	me.cookieSync.Inc()
+}
+
+func (me *Metrics) RecordUserIDSet(userLabels pbsmetrics.UserLabels) {
+	me.userID.With(resolveUserSyncLabels(userLabels)).Inc()
+}
+
+func resolveLabels(labels pbsmetrics.Labels) prometheus.Labels {
+	return prometheus.Labels{
+		"source":  string(labels.Source),
+		"type":    string(labels.RType),
+		"pubid":   labels.PubID,
+		"browser": string(labels.Browser),
+		"cookie":  string(labels.CookieFlag),
+		"status":  string(labels.RequestStatus),
+	}
+}
+
+func resolveAdapterLabels(labels pbsmetrics.AdapterLabels) prometheus.Labels {
+	return prometheus.Labels{
+		"source":  string(labels.Source),
+		"type":    string(labels.RType),
+		"pubid":   labels.PubID,
+		"browser": string(labels.Browser),
+		"cookie":  string(labels.CookieFlag),
+		"status":  string(labels.AdapterStatus),
+		"adapter": string(labels.Adapter),
+	}
+}
+
+func resolveBidLabels(labels pbsmetrics.AdapterLabels, bidType openrtb_ext.BidType, hasAdm bool) prometheus.Labels {
+	bidLabels := prometheus.Labels{
+		"source":  string(labels.Source),
+		"type":    string(labels.RType),
+		"pubid":   labels.PubID,
+		"browser": string(labels.Browser),
+		"cookie":  string(labels.CookieFlag),
+		"status":  string(labels.AdapterStatus),
+		"adapter": string(labels.Adapter),
+		"bidtype": string(bidType),
+		"hasadm":  boolToString(hasAdm),
```

Comment:
`hasadm` would probably be clearer as a `markup_type` label with the values `adm` and `nurl`.

More flexible too. For example, we might want to deprecate `adm` someday and replace it with `html`, `native`, `custom` (in the face of facebook's wonky response JSON), etc.

Reply:
Two problems, one: it is possible that both are present. Two: our current code just checks if `adm` is set, so unless we want to change what we are recording here, this is what we are checking. If we want better `adm`/`nurl` metrics, that should probably be its own PR.